### PR TITLE
Action Queue implementation

### PR DESCRIPTION
**Pull request by DanielChanger**

### CODE REVIEW

File: src/main/java/com/bobocode/bibernate/PersistenceContext.java [hunk 1]
```diff
@@ -25,7 +25,7 @@ public <T> Optional<T> getEntity(Class<T> type, Object key) {
         return Optional.of(type.cast(entity));
     }
 
-    public void putEntity(Object key, Object entity) {
+    public void putEntity(Object entity, Object key) {
```
Comment (nadya-gainutsa) [inline src/main/java/com/bobocode/bibernate/PersistenceContext.java:28]: (nit) why did you change the order of arguments? I liked the prev one, because it was similar to the method put(key, value) in the Map interface.

File: src/main/java/com/bobocode/bibernate/session/SessionImpl.java [hunk 2]
```diff
@@ -109,35 +119,68 @@ public <T> List<T> findAll(Class<T> type, Map<String, Object> properties) {
 
     @Override
     public <T> void save(T entity) {
-        throw new UnsupportedOperationException();
+        Objects.requireNonNull(entity);
+        Validator.validateEntity(entity.getClass());
+        actionQueue.offer(new InsertAction(entityPersister, persistenceContext, entity));
+    }
+
+    public <T> void update(T entity, Map<String, Object> updatedColumns) {
```
Comment (nadya-gainutsa) [inline src/main/java/com/bobocode/bibernate/session/SessionImpl.java]: I guess  this method should be private

**Review by nadya-gainutsa — DISMISSED** [review]

File: src/main/java/com/bobocode/bibernate/session/SessionImpl.java [hunk 3]
```diff
@@ -67,7 +77,7 @@ public <T> Optional<T> find(Class<T> type, Object primaryKey) {
             throw new BibernateException("More than 1 result were found!");
         }
         Optional<T> entity = Optional.of(foundEntities.get(0));
-        persistenceContext.putEntity(primaryKey, entity.get());
+        persistenceContext.putEntity(entity.get(), primaryKey);
```
Comment (StanislavZabramnyi) [inline src/main/java/com/bobocode/bibernate/session/SessionImpl.java]: Maybe lets handle somehow a case where optional is empty to avoid NPE? just calling .get() is not a good practice.
for example some .orElseThrow ->BibernateException(....)

File: src/main/java/com/bobocode/bibernate/session/SessionImpl.java [hunk 4]
```diff
@@ -67,7 +77,7 @@ public <T> Optional<T> find(Class<T> type, Object primaryKey) {
             throw new BibernateException("More than 1 result were found!");
         }
         Optional<T> entity = Optional.of(foundEntities.get(0));
```
Comment (StanislavZabramnyi) [inline src/main/java/com/bobocode/bibernate/session/SessionImpl.java]: What the purpose to wrap this in Optional?  if we have already checked that list is not empty and it contains just 1 element? I guess this value can`t be null.

**Review by StanislavZabramnyi — DISMISSED** [review]
Looks pretty good. I only left two cosmetic comments on the already existing code😬

**Review by StanislavZabramnyi — DISMISSED** [review]

**Review by StanislavZabramnyi — APPROVED** [review]
